docs/source/architecture/data_services_architecture.rst

```diff
@@ -1,4 +1,110 @@
+
+#. Infrastructure Classes, EUDE and DDE: It represents all entities, services, repositories of entities used by MDS Web module to represent MDS entities in UI.
+
+#. Pre Defined Service Section:
+Mds provide some predefined services used by Mds Web and other modules of motech platform :
+
+    *Action Handler Service: It provides methods for handling tasks actions events related
+    with MDS CRUD operations .i.e. create, update and delete.
+
+    * Bundle watcher Suspension Service: An OSGi service, allowing to temporarily disable bundle processing by MDS. When processing gets suspended, MDS will still listen to bundle events, but instead of processing them, they will be queued and processed after the processing gets restored. This allows, besides others, to install/uninstall a larger amount of bundles at one time, without facing annotation processing problems.
+
+    * Csv Export Customizer: It provides custom methods to format related instances during csv export.
+
+    * Csv Import Customizer: It provides custom methods for searching, creating and updating instances during csv import.
+
+    * Csv Import Export Customizer: Service providing methods for importing and exporting instances using csv.
+
+    * Motech Data Service: Centre of Data Access Layer is Generic Data Access Pattern Class which defines generic persistence related methods. Superclass for defining service of any Entity in MDS.
+
+    * Entity Service: It defines CRUD actions on entities.
+
+    * History Service: It persists changes made on given instance. It captures instances which are not in trash.
+
+    * Trash Service: It provides methods related to trash mode (By Default, trash mode is active and can be turned off by user.)
+
+    * Jar Generator Service: It creates a bundle jar with all entities defined in MDS.
+
+    * Json LookUp Service: Service for managing lookups from JSON.
+
+    * MDS Bundle Regeneration Service: It provides methods to regenerate bundle and regenerate bundle after dde enhancement. This service use Jar Generator Service to regenerate bundle.
+
+    * MDS LookUp Service: It provides a methods for processing lookups for particular entity given names, class names as well as String names. It provides a generic access to entities.
+
+    * MDS Scheduler Service: It provides methods to schedule and unschedule a repeating job. It used for emptying trash periodically.
+
+    * Migration Service: It provides method to find flyway migrations within bundle (Default Search location is db/migration).
+
+    * Type Service: It provides methods for defining methods for executing actions on type.
+    Rest Documentation Service. It provides functionality of dynamically generated MDS Rest Documentation.
+
+    * JDO Listener Registry Service: It provides access to registry of Listeners for Persistence events .i.e. POST_CREATE, PRE_DELETE, POST_DELETE, POST_LOAD, PRE_STORE and POST_STORE.It handles pre and post events of crud events of DataNuclues.
```

Comment:
We should describe what this registry does and how (it picks up properly annotated OSGi services and hooks them to JDO lifecycle events)